### PR TITLE
fix: Show correct time until next match

### PR DESCRIPTION
**Pull request by MJ-GH**
Fixes #124 

## Summary of Changes

Properly displays time until next match in the Live Matches column.

## Additional context

Discord username (if different from GitHub): Frozty

## Testing instructions

Sometimes it won't be able to set a time because the json response is not up to date. Usually happens right as a match goes live and the event will show 'state': 'unstarted' when its actually live(should be 'state': 'inProgress'). The issue resolves itself as the data is updated. It gave me a headache during testing, so just keep it in mind.

### Using [GitHub CLI](https://cli.github.com/)
1. Clone this PR
2. Run `gh pr checkout 124` (Requires [GitHub CLI](https://cli.github.com/)) 
3. Follow the [Advanced Installation Guides from the Wiki](https://github.com/LeagueOfPoro/CapsuleFarmerEvolved/wiki)

### Using regular GIT
1. Fetch the PR `git fetch origin pull/<PR_NUMBER>/head:<LOCAL_BRANCH_NAME>` (e.g. `git fetch origin pull/110/head:notif`)
2. Checkout the branch `git checkout <LOCAL_BRANCH_NAME>` (e.g. `git checkout notif`)
3. Follow the [Advanced Installation Guides from the Wiki](https://github.com/LeagueOfPoro/CapsuleFarmerEvolved/wiki)

### CODE REVIEW

**Comment by LeagueOfPoro** [comment]
Can someone test it please?

**Comment by Nub01** [comment]
Seems to be working, I'm in PST, and next stream/game is in 56 mins,

**Comment by MJ-GH** [comment]
[2839e5b](https://github.com/LeagueOfPoro/CapsuleFarmerEvolved/pull/147/commits/2839e5ba58ea62a109a8c29179cb44ca669d50f3) fixes a rare edge case that happened this morning, where some matches were supposed to have started, but they didn't for some reason. My guess is delayed or cancellation. In case it happens again this commit will change the Live matches column to:
billede

**Review by auroraisluna — APPROVED** [review]
LGTM till latest commit. I am unsure if that's going to cause issues. Haven't tested that part either.

**Review by LeagueOfPoro — CHANGES_REQUESTED** [review]
Idea: instead of the weird fix for a delayed match what about changing the message to something like "Up next: LEC at 19:00"

File: src/DataProviderThread.py [hunk 1]
```diff
@@ -85,20 +86,46 @@ def fetchTimeUntilNextMatch(self):
...
-                    hours, remainder = divmod(remainder, 3600)
-                    minutes, seconds = divmod(remainder, 60)
-                    self.sharedData.setTimeUntilNextMatch(f"None - next in {str(days)}d" if days else f'None - next in {hours}h {minutes}m')
+                if event["state"] == "unstarted":
+                    if self._isStartTimeLater(event["startTime"]):  # startTime is later
+                        timeUntil = self.startTime - self.currentTime
+                        total_seconds = int(timeUntil.total_seconds())
+                        days, remainder = divmod(total_seconds, 86400)
+                        hours, remainder = divmod(remainder, 3600)
+                        minutes, seconds = divmod(remainder, 60)
```
[the remark this answers is not in the document]
Comment (MJ-GH) [reply src/DataProviderThread.py]: fixed with commit [388c75f](https://github.com/LeagueOfPoro/CapsuleFarmerEvolved/pull/147/commits/388c75fdee3fd3e4f0106856a73bbbb3547ed5b8)

File: src/DataProviderThread.py [hunk 2]
```diff
@@ -85,20 +86,46 @@ def fetchTimeUntilNextMatch(self):
...
+                        niceStartTime = datetime.strftime(self.startTime, '%H:%M')
+                        self.sharedData.setTimeUntilNextMatch(f"{event['league']['name']} should've started at {niceStartTime} but hasn't.")
+                        for nextEvent in events:  # Looping again to find next match that weren't supposed to have started already
+                            if nextEvent["state"] == "unstarted":
+                                if self._isStartTimeLater(nextEvent["startTime"]):
+                                    timeUntil = self.startTime - self.currentTime
+                                    total_seconds = int(timeUntil.total_seconds())
+                                    days, remainder = divmod(total_seconds, 86400)
+                                    hours, remainder = divmod(remainder, 3600)
+                                    minutes, seconds = divmod(remainder, 60)
```
[the remark this answers is not in the document]
Comment (MJ-GH) [reply src/DataProviderThread.py]: fixed with commit [388c75f](https://github.com/LeagueOfPoro/CapsuleFarmerEvolved/pull/147/commits/388c75fdee3fd3e4f0106856a73bbbb3547ed5b8)

**Comment by MJ-GH** [comment]
> Idea: instead of the weird fix for a delayed match what about changing the message to something like "Up next: LEC at 19:00"

We'd be assuming everyone is in the same timezeone, if we write "19:00", that's why we calculate the time until the next match starts.
So the question is what do you want me to do with the else clause: 
```python
                    else:  # We're past the startTime due to delay or cancellation i'm guessing
                        niceStartTime = datetime.strftime(self.startTime, '%H:%M')
                        self.sharedData.setTimeUntilNextMatch(f"{event['league']['name']} should've started at {niceStartTime} but hasn't.")
                        for nextEvent in events:  # Looping again to find next match that weren't supposed to have started already
                            if nextEvent["state"] == "unstarted":
                                if self._isStartTimeLater(nextEvent["startTime"]):
                                    days, hours, minutes = self._calculateTime()
                                    self.sharedData.addTimeUntilNextMatch(
                                        f"Next in {days}d {hours}h" if days else f'Next in {hours}h {minutes}m')
                                    break
```
Thing is, if the if statement fails and we calculate time, then it will return -1 days. I could just have it say "None - Couldn't determine when next match is", like in my initial commit, or we loop through the events again like it does now, and calculate time once more. And then remove the "LEC should've started at 19:00" part. I can see why it's a weird fix for such a rare case, but it's up to you.

**Comment by Skribb11es** [comment]
Couldn't you just use the system time + the time until the next game to format it correctly according to their timezone? Also shouldn't riot's APIs be returning the timezone-corrected start times anyhow?

**Comment by Nub01** [comment]
> Couldn't you just use the system time + the time until the next game to format it correctly according to their timezone? Also shouldn't riot's APIs be returning the timezone-corrected start times anyhow?

this is what wiflow tried doing isn't it?

**Comment by MJ-GH** [comment]
> Couldn't you just use the system time + the time until the next game to format it correctly according to their timezone? Also shouldn't riot's APIs be returning the timezone-corrected start times anyhow?

Riot's api doesn't do that, the start times are all UTC+0.

**Comment by MJ-GH** [comment]
> Idea: instead of the weird fix for a delayed match what about changing the message to something like "Up next: LEC at 19:00"

commit [c29a502](https://github.com/LeagueOfPoro/CapsuleFarmerEvolved/pull/147/commits/c29a502a81865a45c98432093cb0027641a76df8) makes it print a timezone-correct starttime no matter which timezone you're in, in the format you suggested. 

@Skribb11es showed me on discord how easy that was. Thanks!

I also got rid of the weird fix with the nested for loop. 

I'd like it tested in different timezones than mine(UTC+1) and if you don't want to wait for a real scanario, you can test this commit by adding this line to the end of the try clause between line 104 and 105.
`
print(self.sharedData.getTimeUntilNextMatch())
`
It should print this in the terminal with starttime in your timezone: 
image
Check if the time matches the next match starttime on the schedule here: https://lolesports.com/schedule

**Comment by Skribb11es** [comment]
LGTM, times match up with the correct timestamp for UTC-5, well done.

**Comment by DeadSpc** [comment]
image

help what should i do?

**Comment by DeadSpc** [comment]
image
help

**Comment by auroraisluna** [comment]
> image help

You need to follow the advanced installation guide as the PR description shows, to test it locally.

**Comment by KitsumyFox** [comment]
> image help

I think you just need _**pipenv install**_

**Pull request merged**